apps/dapp/components/dao/DaoTreasuryHistory.tsx

```diff
@@ -43,117 +58,208 @@ export const DaoTreasuryHistory = (props: DaoTreasuryHistoryProps) => {
   )
 }
 
+const NATIVE_DENOM_LABEL = nativeTokenLabel(NATIVE_DENOM)
+// Load roughly 3 days at a time (assuming 1 block per 6 seconds, which is not
+// accurate but close enough).
+const BLOCK_HEIGHT_INTERVAL = (60 * 60 * 24 * 3) / 6
+
 export const InnerDaoTreasuryHistory = ({
   shortTitle,
 }: DaoTreasuryHistoryProps) => {
   const { t } = useTranslation()
   const { coreAddress } = useDaoInfoContext()
-  const transactions = useRecoilValue(treasuryTransactionsSelector(coreAddress))
 
-  return transactions.length ? (
-    <div className="space-y-4">
+  // Initialization.
+  const latestBlockHeight = useRecoilValue(blockHeightSelector)
+  const initialMinHeight = latestBlockHeight - BLOCK_HEIGHT_INTERVAL
+  const initialLowestHeightLoadedTimestamp = useRecoilValue(
+    blockHeightTimestampSafeSelector(initialMinHeight)
+  )
+  const initialTransactions = useRecoilValue(
+    transformedTreasuryTransactionsSelector({
+      address: coreAddress,
+      minHeight: initialMinHeight,
+      maxHeight: latestBlockHeight,
+    })
+  )
+
+  // Paginated data.
+  const [loading, setLoading] = useState(false)
+  const [lowestHeightLoaded, setLowestHeightLoaded] = useState(initialMinHeight)
+  const [lowestHeightLoadedTimestamp, setLowestHeightLoadedTimestamp] =
+    useState(initialLowestHeightLoadedTimestamp)
+  const [transactions, setTransactions] = useState(initialTransactions)
+  const [canLoadMore, setCanLoadMore] = useState(true)
+
+  // Pagination loader.
+  const loadMoreTransactions = useRecoilCallback(
+    ({ snapshot }) =>
+      async (maxHeight: number) => {
+        const minHeight = maxHeight - BLOCK_HEIGHT_INTERVAL
+
+        setLoading(true)
+        try {
+          const newTransactions = await snapshot.getPromise(
+            transformedTreasuryTransactionsSelector({
+              address: coreAddress,
+              minHeight,
+              maxHeight,
+            })
+          )
+
+          const newLowestHeightLoadedTimestamp = await snapshot.getPromise(
+            blockHeightTimestampSelector(minHeight)
+          )
+
+          setLowestHeightLoaded(minHeight)
+          setLowestHeightLoadedTimestamp(newLowestHeightLoadedTimestamp)
+
+          // If no transactions found, try to load more.
+          if (!newTransactions.length) {
+            return await loadMoreTransactions(minHeight)
+          }
+
+          setTransactions((transactions) => [
+            ...transactions,
+            ...newTransactions,
+          ])
+        } catch (err) {
+          console.error(
+            processError(err, { tags: { coreAddress, minHeight, maxHeight } })
+          )
+          // If errored, assume we cannot load any more below this height.
+          setCanLoadMore(false)
+        } finally {
+          setLoading(false)
+        }
+      },
+    [
+      coreAddress,
+      setLoading,
+      setTransactions,
+      setLowestHeightLoaded,
+      setLowestHeightLoadedTimestamp,
```

Comment:
might wanna add `BLOCK_HEIGHT_INTERVAL` to this array, but should be fine without